mate-panel/panel-run-dialog.c

```diff
@@ -766,79 +776,83 @@ static GSList *
 get_all_applications_from_alias (MateMenuTreeAlias *alias,
 				 GSList         *list)
 {
-	MateMenuTreeItem *aliased_item;
-
-	aliased_item = matemenu_tree_alias_get_item (alias);
-
-	switch (matemenu_tree_item_get_type (aliased_item)) {
-	case MATEMENU_TREE_ITEM_ENTRY:
-		list = g_slist_append (list, matemenu_tree_item_ref (aliased_item));
-		break;
-
-	case MATEMENU_TREE_ITEM_DIRECTORY:
-		list = get_all_applications_from_dir (MATEMENU_TREE_DIRECTORY (aliased_item),
-						      list);
-		break;
-
-	default:
-		break;
+	gpointer item;
+	switch (matemenu_tree_alias_get_aliased_item_type (alias)) {
+		case MATEMENU_TREE_ITEM_ENTRY:
+			item = matemenu_tree_alias_get_aliased_entry (alias);
+			list = g_slist_append (list, (MateMenuTreeEntry*) item);
+			break;
+		case MATEMENU_TREE_ITEM_DIRECTORY:
+			item = matemenu_tree_alias_get_aliased_directory (alias);
+			list = get_all_applications_from_dir ((MateMenuTreeDirectory*) item, list);
+			break;
+		default:
+			break;
 	}
-
-	matemenu_tree_item_unref (aliased_item);
-
 	return list;
 }
 
 static GSList *
 get_all_applications_from_dir (MateMenuTreeDirectory *directory,
 			       GSList             *list)
 {
-	GSList *items;
-	GSList *l;
-
-	items = matemenu_tree_directory_get_contents (directory);
-
-	for (l = items; l; l = l->next) {
-		switch (matemenu_tree_item_get_type (l->data)) {
-		case MATEMENU_TREE_ITEM_ENTRY:
-			list = g_slist_append (list, matemenu_tree_item_ref (l->data));
-			break;
-
-		case MATEMENU_TREE_ITEM_DIRECTORY:
-			list = get_all_applications_from_dir (l->data, list);
-			break;
-
-		case MATEMENU_TREE_ITEM_ALIAS:
-			list = get_all_applications_from_alias (l->data, list);
-			break;
-
-		default:
-			break;
+	MateMenuTreeIter *iter;
+	MateMenuTreeItemType type;
+
+	iter = matemenu_tree_directory_iter (directory);
+	while ((type = matemenu_tree_iter_next (iter)) != MATEMENU_TREE_ITEM_INVALID) {
+		gpointer item;
+		switch (type) {
+			case MATEMENU_TREE_ITEM_ENTRY:
+				item = matemenu_tree_iter_get_entry (iter);
```

Comment:
In other places, similarly created `item` is later unref'ed via `matemenu_tree_item_unref`. Looks like it's missing here (in all three cases).

Reply:
Ok, the item is left with reference if it's added to list with `g_slist_append`. But what if it's a directory? Shouldn't we unref it after `get_all_applications_from_dir` call is finished?

Reply:
Seems like the same applies to `get_all_applications_from_alias` call as well.

Reply:
If call matemenu_tree_item_unref here, the running dialog(Alt-F2) will crash.

Reply:
Not right here, but after recursive calls to `get_all_applications_from_dir` and `get_all_applications_from_alias`. They don't seem to add their `directory` and `alias` arguments to any list, so these can be unref'ed after the function call is finished.

Reply:
Quickly looking `iter` is created before `while` loop and unrefed after it. But I guess you think that `item` should be unrefed, right?

At least in GNOME `gmenu_tree_iter_get_entry` has `Returns: (transfer full): An entry` and if same thing is in MATE then it looks like there are memory leak...

I think you must make sure that all items that are added to list should be freed! If somewhere some functions returns `transfer none` then you should ref it before adding to list.

And then when you are done with list, you should free it with `g_slist_free_full (all_applications, matemenu_tree_item_unref);`.

Reply:
Please see [here](https://github.com/mate-desktop/mate-panel/pull/794/files#diff-065586010f2f4c80630c35dce7531240R923), already use g_slist_free_full to free them.
```
-	g_slist_free (all_applications);
+	g_slist_free_full (all_applications, matemenu_tree_item_unref);
```
If do that like @monsta said, call  `matemenu_tree_item_unref` here, Alt-F2 will crash at `g_slist_free_full(all_applications, matemenu_tree_item_unref)` line.

Reply:
Oh, I missed that, but I am quite sure I saw `g_slist_free (all_applications);`... If it crash in `g_slist_free_full` then you have some item in list that is missing item_ref call (because it returned from function with transfer none) before it is added to list... Or it is unrefed when it should not... Or no idea... Maybe post stacktrace of crash?

Reply:
I didn't mean the item that's added to the list via `g_slist_append`. I meant the one that is passed into `get_all_applications_from_dir` and `get_all_applications_from_alias`.

Reply:
@monsta is right, you need to unref item that is returned from `matemenu_tree_iter_get_directory` and `matemenu_tree_iter_get_alias`.

Reply:
@monsta You are right, the code is update, please review, thank you.

Reply:
Thanks, now the memory should be freed properly.